### PR TITLE
feature: Alert banner for missing data

### LINKED ISSUE #288: Warning banner for missing data
## Use Case

Implement a warning banner that appears once frames are found to be missing in the dataset.

Image

Needs to have:
- Checkbox, "don't show again for this dataset"
- Title prop
- Additional info prop
- "Read more" fold text/button

## Details

- Add an event listener in the CanvasWrapper that can pass up to a higher component? Or for a cleaner option, use Context to pass it down?
- Create a new component for the banner

### PR DESCRIPTION
**Pull request by ShrimpCryptid**
Problem
=======
*Estimated review size: medium, 30 minutes*

Closes #288, implementing a warning banner that appears when data is missing.

Solution
========
Oops, this one got a little bigger than I intended! I was trying to make the alert banner system really robust (which. it maybe didn't need to be.)

- Adds the new `AlertBanner` component, which is a wrapper around Ant's Alert component.
  - It enforces certain styling rules (like the `banner` flag) and modifies some of the styling to match UX designs.
  - `AlertBanner` can show a "Don't show again for this dataset" checkbox, the value of which is reported when the banner is closed.
- Adds the `useAlertBanner` hook, which returns a ReactElement list of banners to render, and a callable method for adding new banners.
  - `useAlertBanner` only allows exactly one banner per message, and handles the logic for the "Don't show again" checkbox.
  - Banners will be reset when the dependencies (usually `dataset`) is reset.

Other changes:
- Alert banner is now shown when a bad URL is loaded. (see below verification steps!)
- Adds underlining to focused links for better accessibility :)

## Type of change
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------

Testing error banner:
1. Open preview link. This should fail to load the dataset, and will show the error banner: https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-321/#/viewer?collection=bad

Testing repeated warning:
1. Open preview link. Start playback. https://allen-cell-animated.github.io/nucmorph-colorizer/pr-preview/pr-321/#/viewer?collection=https%3A%2F%2Fdev-aics-dtp-001.int.allencell.org%2Fassay-dev%2Fcomputational%2Fdata%2FEMT_deliverable_processing%2FTranscription_factor_cell_lines%2FVersion_1%2Fcolorizer_output%2FEOMES%2Fcollections_all.json&dataset=AICS-122_3D-MG-EMT-no-MG_fms_id_592af19bfc414ae1bf05b81967190944
2. Around frame 26, a warning banner should appear.
3. Close the banner and start playback again from frame 20. At frame 26, the same banner should appear. Playing past 26 again will not cause another banner to appear.
4. Check the "Do not show again" checkbox and close the alert. The alert will not reappear.
5. Switch to another dataset. The banner should disappear.

Screenshots (optional):
-----------------------

**Video talk-through (🔊):**

https://github.com/allen-cell-animated/nucmorph-colorizer/assets/30200665/875d9c1f-1b4c-4406-b0ae-1118f4ffe184

image
image

Keyfiles (delete if not relevant):
-----------------------
1. `AlertBanner.tsx`
6. `hooks.tsx`: Contains a hook for queueing up alert banners.

### CODE REVIEW

File: src/Viewer.tsx [hunk 1]
```diff
@@ -458,11 +461,14 @@ function Viewer(): ReactElement {
             datasetKey = datasetParam || newCollection.getDefaultDatasetKey();
           } catch (error) {
             console.error(error);
-            notificationApi["error"]({
-              message: "Error loading dataset:",
-              description: (error as Error).message,
-              placement: "bottomLeft",
-              duration: 4,
```
Comment (ShrimpCryptid) [inline src/Viewer.tsx:465]: This error message flashes pretty quickly, so I decided to replace it with the new alert banner. It includes more helpful context and an immediate link for users to return to the homepage!

File: src/components/AppStyle.tsx [hunk 2]
```diff
@@ -219,6 +237,13 @@ const CssContainer = styled.div`
   margin: 0;
   color: var(--color-text-primary);
 
+  a {
+    &:focus,
+    &:focus-visible {
+      text-decoration: underline;
+    }
+  }
+
```
Comment (ShrimpCryptid) [inline src/components/AppStyle.tsx:246]: Misc change. Makes all links show an underline when focused by keyboard navigation.

File: src/colorizer/ColorizeCanvas.ts
Comment (ShrimpCryptid) [inline src/colorizer/ColorizeCanvas.ts:1]: Renamed the callback for clarity!

File: src/components/Banner/AlertBanner.tsx [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+  }
+`;
+
+export type AlertBannerProps = Spread<
+  Omit<AlertProps, "onClose" | "afterClose" | "message" | "description" | "closable" | "banner"> & {
+    message: string;
+    /** Additional text, hidden behind a button labeled "Read more". Use an array for multiple lines.*/
+    description?: string | string[];
+    /** If true, will show a checkbox reading, "Do not show again for this dataset." */
+    showDoNotShowAgainCheckbox?: boolean;
```
Comment (ShrimpCryptid) [inline src/components/Banner/AlertBanner.tsx:67]: I'm accepting better names for this 😅

**Comment by ShrimpCryptid** [comment]
Taking this back to draft because there are some UX suggestions I need to make! The core functionality will all stay the same though.

File: src/components/Banner/AlertBanner.tsx (hunk 3 above)
Comment (toloudis) [reply src/components/Banner/AlertBanner.tsx:77]: is good!

File: src/components/Banner/AlertBanner.tsx [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+/**
+ * A banner-style alert that wraps around the Ant Alert component.
+ * @param type: Type of alert to show. Can be "info", "warning" (default), "error", or "success".
+ * @param message: The main message to display in the alert.
+ * @param description: The description to display in the alert. This will be hidden behind a "Read more" button until clicked.
+ * @param showDoNotShowAgainCheckbox: If true, will show a checkbox reading, "Do not show again for this dataset."
+ * @param action: A button or other component element to show in the alert.
+ *
+ * Please consider using the `useAlertBanner` hook to manage alert banner states!
```
Comment (toloudis) [inline src/components/Banner/AlertBanner.tsx:91]: consider, or definitely use?

File: src/components/CanvasWrapper.tsx [hunk 5]
```diff
@@ -86,9 +89,24 @@ export default function CanvasWrapper(inputProps: CanvasWrapperProps): ReactElem
...
+            "Check your network connection and access to the dataset path, or use the browser console to view details. Otherwise, contact the dataset creator as there may be missing files.",
+          showDoNotShowAgainCheckbox: true,
+          closable: true,
+        });
+      }
+    },
+    [props.showAlert, setShowMissingFileIcon, canv]
+  );
+
+  canv.setOnFrameChangeCallback(onFrameChangedCallback);
```
Comment (toloudis) [inline src/components/CanvasWrapper.tsx:109]: I'm a little unsure of this diff. Previously this was being called inside a useEffect.  Is the above useCallback sort of replacing that? Could you still be using useEffect here?

**Review by toloudis — APPROVED** [review]

File: src/components/CanvasWrapper.tsx (hunk 5 above)
Comment (ShrimpCryptid) [reply src/components/CanvasWrapper.tsx:109]: I probably didn't need to be using `useEffect` in the first place!

`useCallback` caches the definition of `onFrameChangedCallback` unless its dependencies change, which I'm using as a minor optimization.

`canv.setOnFrameChangeCallback` doesn't really incur any costs, it just causes `canv` to set an internal field, so there's no real concern with running it on each update. (I could also cache this through a `useMemo` but again, this should be fine.)

File: src/components/Banner/AlertBanner.tsx (hunk 4 above)
Comment (ShrimpCryptid) [reply src/components/Banner/AlertBanner.tsx:91]: `useAlertBanner` is useful for dynamically adding/removing banners, but if we want a banner to be visible from initial load it makes more sense to directly add the `AlertBanner` component. I'll clarify what this means in the comment.

File: src/components/Banner/hooks.tsx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+  // banner or only ever show the first call?
+  // TODO: Nice animations when banners appear or when all of them are cleared at once?
+  const [bannerProps, setBannerProps] = useReducer(
+    (_currentBanners: AlertBannerProps[], newBanners: AlertBannerProps[]) => newBanners,
+    []
+  );
+  const ignoredBannerMessages = useRef(new Set<string>());
+
+  const showAlert = useCallback(
+    // Modify props to add a listener for the close event so we can delete the banner.
```
Comment (frasercl) [inline src/components/Banner/hooks.tsx:49]: A bit confused by this comment. Is it describing the behavior of the `map` used to create `bannerElements` below?

File: src/Viewer.tsx [hunk 7]
```diff
@@ -146,6 +147,8 @@ function Viewer(): ReactElement {
   };
   const [notificationApi, notificationContextHolder] = notification.useNotification(notificationConfig);
 
+  const { bannerElement: bannerEl, showAlert } = useAlertBanner([dataset]);
```
Comment (frasercl) [inline src/Viewer.tsx:150]: teeny optional nit: not getting a lot out of renaming `bannerElement` when destructuring here

File: src/components/Banner/AlertBanner.tsx [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    align-items: flex-start;
+    flex-wrap: wrap;
+    padding: 8px 30px;
+
+    ${(props) => {
+      return css`
+        border-bottom: 1px solid var(--color-alert-${props.type}-border) !important;
+      `;
+    }}
+  }
```
[the remark this answers is not in the document]
Comment (ShrimpCryptid) [reply src/components/Banner/AlertBanner.tsx:23]: Oop, it totally can.

File: src/components/Banner/AlertBanner.tsx [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        )}
+      </FlexRowAlignCenter>
+      {showFullContent && <FlexColumn>{description}</FlexColumn>}
+    </FlexColumn>
+  );
+
+  newProps.message = message;
+
+  // Override the "message" prop on the Alert with a custom react element
+  return <StyledAlert {...newProps}></StyledAlert>;
```
[the remark this answers is not in the document]
Comment (ShrimpCryptid) [reply src/components/Banner/AlertBanner.tsx:155]: Nice catch, thank you :)

**Pull request merged**